core/src/main/java/io/stargate/core/activator/BaseActivator.java

```diff
@@ -0,0 +1,169 @@
+    for (Class<?> dependentService : dependentServices) {
+      map.put(dependentService, null);
+    }
+    return map;
+  }
+
+  /**
+   * When the OSGi invokes the start() method, it will try to get all dependentServices passed to
+   * this class's constructor. It will use the {@link BundleContext#getServiceReference(String)}. If
+   * any of those services are not present, it will register the listeners for them using {@link
+   * BundleContext#addServiceListener(ServiceListener, String)} and not start the target service
+   * (see targetServiceClass). It will wait for a notification denoting that service was registered
+   * using {@link this#serviceChanged(ServiceEvent)}. If all services are present, it will call the
+   * user-provided {@link this#createService(List)} and register it in the OSGi using {@link
+   * BundleContext#registerService(Class, Object, java.util.Dictionary)}.
+   */
+  @Override
+  public synchronized void start(BundleContext context) throws InvalidSyntaxException {
+    logger.info("Starting {} ...", activatorName);
+    this.context = context;
+
+    for (Class<?> dependentService : dependentServices) {
+      ServiceReference<?> serviceReference =
+          context.getServiceReference(dependentService.getName());
+      if (serviceReference == null) {
+        logger.debug(
+            "{} service is null, registering a listener to get notification when it will be ready.",
+            dependentService.getName());
+        context.addServiceListener(
+            this, String.format("(objectClass=%s)", dependentService.getName()));
+      } else {
+        // if the service is present, get it
+        registeredServices.put(dependentService, context.getService(serviceReference));
+      }
+    }
+
+    if (registeredServices.values().stream().anyMatch(Objects::isNull)) {
+      // It will be started once all dependent services are registered
+      return;
+    }
+
+    // all dependant services are present
+    startServiceInternal(new LinkedList<>(registeredServices.values()));
+  }
+
+  @Override
+  public synchronized void stop(BundleContext context) throws Exception {
+    // no-op
+  }
+
+  /**
+   * It will try to match all dependentServices with a ServiceEvent notification. If the
+   * notification for dependant service contains a non-null value, it will start the target service.
+   * After the notification is handled, it checks if all dependentServices are not null. If they
+   * are, the client's provided {@link this#createService(List)} is called, and the service is
+   * registered. It will not register the service if it was already registered in the {@link
+   * this#start(BundleContext)}.
+   */
+  @Override
+  public synchronized void serviceChanged(ServiceEvent event) {
+    for (Class<?> dependentService : dependentServices) {
+      Object registeredService = BundleUtils.getRegisteredService(context, event, dependentService);
```

Comment:
what about dependent service de-registration?

Reply:
I've reworked this example to use `Tracker`.
It already provides the way to unget service:
```
	public void removedService(ServiceReference<S> reference, T service) {
		context.ungetService(reference);
	}
```
So I think that we don't need to handle it explicitly.